### PR TITLE
Fix/tao 4300 prevent print screen

### PR DESCRIPTION
**Pull request by eacraft13**
[tao-4300](https://oat-sa.atlassian.net/browse/TAO-4300)

Companion [extension-tao-act pull request](https://github.com/oat-sa/extension-tao-act/pull/542)

Due to print screen / screenshot being an OS level command, it is not possible to capture in the browser.

For mac, Cmd+Shift+[3|4] cannot be captured on keyup or keydown. So, to prevent screenshots, a blur is added on keydown of Cmd+Shift. And removed on keyup of any key when Cmd+Shift is still not pressed. This does have a bit of a shortcoming when Cmd+Shift+4 is used. To remove blur, the user will have to press any key that is not Cmd+Shift.

For pc, the solution is different due to the nature of PrtScn. PrtScn cannot be captured on keydown, so an unblurred screenshot cannot be prevented. But, PrtScn can be captured on keyup. So, on keyup of PrtScn, I chose to pause execution so that the proctor would be notified that a successful screenshot was taken.

### CODE REVIEW

File: views/js/runner/plugins/security/preventScreenshot.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                    }
+                });
+            }
+
+            // Windows - pause on PrtScn
+            else if (platform === 'win') {
+                $(window)
+                .on('keyup', function (e) {
+                    if (e.keyCode === 44) {
+                        testRunner
```
Comment (no-chris) [inline views/js/runner/plugins/security/preventScreenshot.js]: You can enforce the prohibited behavior reporting by triggering also this event
`testRunner.trigger('prohibited-key', 'printScreen usage');`

File: views/js/runner/plugins/security/preventScreenshot.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+         * Called when the host is installing the plugins
+         * @returns {Promise} to resolve async delegation
+         */
+        install: function install() {
+            var testRunner = this.getTestRunner();
+
+            // For mac - blur on Cmd+Shift
+            if (platform === 'mac') {
+                $(window)
+                .on('keydown', function (e) {
```
Comment (no-chris) [inline views/js/runner/plugins/security/preventScreenshot.js]: This event should be namespaced
`.on('keydown.preventScreenshot', function (e) {`

File: views/js/runner/plugins/security/preventScreenshot.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    if (e.metaKey && e.shiftKey) {
+                        $('body').css('filter', 'blur(20px)');
+                    }
+                })
+                // Note - When user hits Cmd+Shift+4, they must press any key
+                // to remove blur (that is not Cmd+Shift)
+                .on('keyup', function (e) {
```
Comment (no-chris) [inline views/js/runner/plugins/security/preventScreenshot.js]: same here

File: views/js/runner/plugins/security/preventScreenshot.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                                category: __('examinee'),
+                                subCategory: __('behaviour')
+                            }
+                        });
+                    }
+                });
+            }
+
+            testRunner.on('destroy', function () {
+                $(window).off('keydown keyup');
```
Comment (no-chris) [inline views/js/runner/plugins/security/preventScreenshot.js]: Simply remove the namespace so you do not accidently unbind other handlers
`$(window).off('.preventScreenshot');`

File: views/js/runner/plugins/security/preventScreenshot.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                    if (!e.metaKey || !e.shiftKey) {
+                        $('body').css('filter', '');
+                    }
+                });
+            }
+
+            // Windows - pause on PrtScn
+            else if (platform === 'win') {
+                $(window)
+                .on('keyup', function (e) {
```
Comment (no-chris) [inline views/js/runner/plugins/security/preventScreenshot.js]: same here

File: views/js/runner/plugins/security/preventScreenshot.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2016 (original work) Open Assessment Technologies SA ;
```
Comment (no-chris) [inline views/js/runner/plugins/security/preventScreenshot.js]: copyright year

File: views/js/runner/plugins/security/preventScreenshot.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    'use strict';
+
+    /**
+     * Identify the current platform
+     * @type {String}
+     */
+    var platform = navigator.platform.indexOf('Mac') < 0 ? 'win' : 'mac';
+
+    /**
+     * Creates the preventCopy plugin.
```
Comment (no-chris) [inline views/js/runner/plugins/security/preventScreenshot.js]: copy/paste leftovers ;-)

File: views/js/runner/plugins/security/preventScreenshot.js (hunk 1 above)
Comment (eacraft13) [inline views/js/runner/plugins/security/preventScreenshot.js]: To be clear, triggering 'prohibited-key' should be in addition to triggering a 'pause', correct?

**Pull request merged**